kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -117,307 +134,120 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
-                is OpDescriptor -> state.perform(this)
-                is LockedQueue -> return "Mutex[${state.owner}]"
-                else -> error("Illegal state $state")
-            }
-        }
-    }
+    @Suppress("UNCHECKED_CAST")
+    override val onLock: SelectClause2<Any?, Mutex> get() = SelectClause2Impl(
+        clauseObject = this,
+        regFunc = MutexImpl::onLockRegFunction as RegistrationFunction,
+        processResFunc = MutexImpl::onLockProcessResult as ProcessResultFunction,
+        onCancellationConstructor = onCancellationUnlockConstructor
+    )
 
-    private class LockedQueue(
-        @JvmField var owner: Any
-    ) : LockFreeLinkedListHead() {
-        override fun toString(): String = "LockedQueue[$owner]"
+    protected open fun onLockRegFunction(select: SelectInstance<*>, owner: Any?) {
+        onAcquire.regFunc(this, SelectInstanceWithOwner(select, owner), owner)
     }
 
-    private abstract inner class LockWaiter(
-        @JvmField val owner: Any?
-    ) : LockFreeLinkedListNode(), DisposableHandle {
-        private val isTaken = atomic(false)
-        fun take(): Boolean = isTaken.compareAndSet(false, true)
-        final override fun dispose() { remove() }
-        abstract fun tryResumeLockWaiter(): Boolean
-        abstract fun completeResumeLockWaiter()
+    protected open fun onLockProcessResult(owner: Any?, result: Any?): Any? {
+        onAcquire.processResFunc(this, null, result)
+        return this
     }
 
-    private inner class LockCont(
-        owner: Any?,
-        private val cont: CancellableContinuation<Unit>
-    ) : LockWaiter(owner) {
+    protected val onCancellationUnlockConstructor: OnCancellationConstructor =
+        { select: SelectInstance<*>, owner: Any?, ignoredResult: Any? ->
+            { unlock(owner) }
+        }
 
-        override fun tryResumeLockWaiter(): Boolean {
-            if (!take()) return false
-            return cont.tryResume(Unit, idempotent = null) {
-                // if this continuation gets cancelled during dispatch to the caller, then release the lock
-                unlock(owner)
-            } != null
+    private inner class CancellableContinuationWithOwner(
+        val cont: CancellableContinuation<Unit>,
+        val owner: Any?
+    ) : CancellableContinuation<Unit> by cont {
+        override fun tryResume(value: Unit, idempotent: Any?, onCancellation: ((cause: Throwable) -> Unit)?): Any? {
+            val token = cont.tryResume(value, idempotent, onCancellation)
+            if (token !== null) this@MutexImpl.owner.value = owner
+            return token
         }
 
-        override fun completeResumeLockWaiter() = cont.completeResume(RESUME_TOKEN)
-        override fun toString(): String = "LockCont[$owner, ${cont}] for ${this@MutexImpl}"
+        override fun resume(value: Unit, onCancellation: ((cause: Throwable) -> Unit)?) {
+            this@MutexImpl.owner.value = owner
+            cont.resume(value, onCancellation)
```

Comment:
Here it seems like it is possible to have an unlocked mutex with an owner different from `NO_OWNER` in case the cancellation of the outer job (that is parent of current `CancellableContinuationWithOwner`) happens right between setting owner and calling `resume`.

Reply:
Nice catch! Indeed, an `onCancellation` action should be provided to this `resume` call.

Reply:
Surprisingly, the current implementation works _almost_ correctly. After this buggy situation, the following lock acquisition that successfully acquires the semaphore permit _blindly_ overwrites owner. I have fixed this behavior, adding more invariant `assert`-s to the algorithm and a new stress test.